core/validation/casting.py

```diff
@@ -18,24 +20,23 @@ def cast_to_schema(data: dict[str, pd.DataFrame], schema: dict) -> None:
             continue  # skip casting if schema doesn't exist for that table_data - this will be caught during validation
 
         column_to_type = schema[table]["columns"]
-        table_types = table_data.dtypes
-
-        table_retyped = False
-        for column, target_type in column_to_type.items():
-            if column in table_types:
-                original_type = table_types[column]
-
-                if original_type != target_type:
-                    if target_type == "list":
-                        # Forcible conversion of values to target_type == "list" occurs in extract_postcodes()
-                        # and .astype does not support lists
-                        continue
-
-                    try:
-                        table_data[column] = table_data[column].astype(target_type)
-                        table_retyped = True
-                    except ValueError:
-                        continue  # this will be caught during validation
-
-        if table_retyped:
-            data[table] = table_data
+
+        for pos, (index, row) in enumerate(table_data.iterrows()):
+            for column, value in row.items():
+                if pd.isna(value) or isinstance(value, (datetime, pd.Timestamp)):
+                    continue  # do not cast nan or datetime
+
+                try:
+                    table_data.iloc[pos, table_data.columns.get_loc(column)] = column_to_type[column](value)
+                except (TypeError, ValueError):
+                    continue  # if we can't cast, leave for validation to catch
+
+
+_NUMPY_TO_PY_TYPES = {
+    "object": "object",
+    "int64": int,
+    "bool": bool,
+    "float64": float,
+    "datetime64[ns]": datetime,
+    "<M8[ns]": datetime,
+}
```

Comment:
What are we now using this for?

Reply:
@cyrusdobbs I think it's because we were still having issues with multiple values being cast as objects, and a mixture of datetime/timestamp types when we just removed the mapping completely - does that sound accurate to you?

Reply:
Looking at the code this isn't actually used anywhere now.. so we can delete it